src/GoogleSignin.ios.js

```diff
@@ -1,157 +1,143 @@
-    );
+      <RNGoogleSigninButton
+        style={[{ backgroundColor: 'rgba(0,0,0,0)' }, style]}
+        onPress={this.handleOnPress}
+        {...props}
+      />
+    )
   }
 }
 
 GoogleSigninButton.Size = {
   Icon: RNGoogleSignin.BUTTON_SIZE_ICON,
   Standard: RNGoogleSignin.BUTTON_SIZE_STANDARD,
   Wide: RNGoogleSignin.BUTTON_SIZE_WIDE
-};
+}
 
 GoogleSigninButton.Color = {
   Auto: RNGoogleSignin.BUTTON_COLOR_AUTO,
   Light: RNGoogleSignin.BUTTON_COLOR_LIGHT,
   Dark: RNGoogleSignin.BUTTON_COLOR_DARK
-};
+}
 
 class GoogleSignin {
-
-  constructor() {
-    this._user = null;
-    this.signinIsInProcess = false;
-  }
+  _user = null
+  signinIsInProcess = false
 
   hasPlayServices(params = { autoResolve: true }) {
-    return Promise.resolve(true);
+    return Promise.resolve(true)
   }
 
-  configure(params={}) {
+  configure(params = {}) {
     if (!params.iosClientId) {
-      throw new Error('GoogleSignin - Missing iOS app ClientID');
+      return Promise.reject(new Error('RNGoogleSignin: Missing iOS app ClientID'))
     }
 
     if (params.offlineAccess && !params.webClientId) {
-      throw new Error('GoogleSignin - offline use requires server web ClientID');
+      return Promise.reject(new Error('RNGoogleSignin: offline use requires server web ClientID'))
     }
 
-    params = [
-	params.scopes || [], params.iosClientId, params.offlineAccess ? params.webClientId : '', params.hostedDomain ? params.hostedDomain : null
-    ];
+    const config = [
+      params.scopes || [],
+      params.iosClientId,
+      params.offlineAccess ? params.webClientId : '',
+      params.hostedDomain ? params.hostedDomain : null
+    ]
 
-    RNGoogleSignin.configure(...params);
-    return Promise.resolve(true);
+    return RNGoogleSignin.configure(...config)
   }
 
   currentUserAsync() {
-    return new Promise((resolve, reject) => {
-      const sucessCb = NativeAppEventEmitter.addListener('RNGoogleSignInSuccess', (user) => {
-        this._user = {
-          ...user,
-        };
-        this._removeListeners(sucessCb, errorCb);
-        resolve(user);
-      });
-
-      const errorCb = NativeAppEventEmitter.addListener('RNGoogleSignInError', () => {
-        this._removeListeners(sucessCb, errorCb);
-        resolve(null);
-      });
-
-      RNGoogleSignin.currentUserAsync();
-    });
+    return RNGoogleSignin.currentUserAsync()
+      .then(user => {
+        this._user = { ...user }
+        return user
+      })
+      .catch(error => {
+        // The user has never signed in before with the given scopes, or they have since signed out.
+        if (error.code === '-4') {
+          this._user = null
+          this.signinIsInProcess = false
+          return null
+        }
+
+        this.signinIsInProcess = false
+        throw error
+      })
   }
 
   currentUser() {
-    return {...this._user};
+    if (this._user) {
+      return { ...this._user }
```

Comment:
I think the fact we're always returning a new object deserves a comment (why?)

Reply:
I dont personally need any explanation since treating everything as immutable is quite popular in React world. More specifically returning a new object is necessary since “we don’t want users to mess up our library internals when they make modifications to the user object that is returned”.

Quoted sentence would be my comment. :)

Reply:
> treating everything as immutable

this also means that you return a new object only if it is different from the previous one. Here, if you call

```
const u1 = ourModule.currentUser()
const u2 = ourModule.currentUser()

u1 === u2 // this is false, which I don't think is expected
```

> we don’t want users to mess up our library internals when they make modifications to the user object that is returned

I do agree with that, but I'd also argue that writing to the object returned would be a pretty unclever idea. Then again, I guess some people would do that...

I think I saw a PR with such changes, let me find out what the motivation was

Reply:
hum, [this](https://github.com/react-native-community/react-native-google-signin/commit/8364228fb6cfdff2fd99695e174cd9cd0dde39f6) is similar but different.. and no context. So let's just leave it.